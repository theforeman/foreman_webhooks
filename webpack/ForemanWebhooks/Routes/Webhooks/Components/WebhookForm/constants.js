export const HTTP_METHODS = [
  { value: 'POST', label: 'POST' },
  { value: 'GET', label: 'GET' },
  { value: 'PUT', label: 'PUT' },
  { value: 'DELETE', label: 'DELETE' },
  { value: 'PATCH', label: 'PATCH' },
];

export const WEBHOOK_ITEM = 'Webhook';
