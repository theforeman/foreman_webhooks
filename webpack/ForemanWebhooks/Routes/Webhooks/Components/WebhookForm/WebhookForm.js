import React, { useState } from 'react';
import PropTypes from 'prop-types';
import * as Yup from 'yup';

import { translate as __ } from 'foremanReact/common/I18n';
import ForemanForm from 'foremanReact/components/common/forms/ForemanForm';

import WebhookFormTabs from './Components/WebhookFormTabs';

import { HTTP_METHODS, WEBHOOK_ITEM } from './constants';

const webhookFormSchema = Yup.object().shape({
  name: Yup.string().required(__('is required')),
  target_url: Yup.string().required(__('is required')),
  http_method: Yup.string().required(__('is required')),
  event: Yup.string().required(__('is required')),
  webhook_template_id: Yup.string().required(__('is required')),
});

const WebhookForm = ({
  onCancel,
  handleSubmit,
  initialValues,
  templates,
  availableEvents,
  isTemplatesLoading,
  isEventsLoading,
}) => {
  const webhookTemplates = templates.map(t => ({ value: t.id, label: t.name }));

  const [activeTab, setActiveTab] = useState(0);
  const handleTabClick = (event, tabIndex) => {
    event.preventDefault();
    setActiveTab(tabIndex);
  };

  return (
    <ForemanForm
      onSubmit={handleSubmit}
      initialValues={initialValues}
      validationSchema={webhookFormSchema}
      onCancel={onCancel}
      enableReinitialize
      item={WEBHOOK_ITEM}
    >
      <WebhookFormTabs
        activeTab={activeTab}
        handleTabClick={handleTabClick}
        webhookTemplates={webhookTemplates}
        httpMethods={HTTP_METHODS}
        availableEvents={availableEvents}
        isEventsLoading={isEventsLoading}
        isTemplatesLoading={isTemplatesLoading}
      />
    </ForemanForm>
  );
};

WebhookForm.propTypes = {
  onCancel: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.object.isRequired,
  templates: PropTypes.array.isRequired,
  availableEvents: PropTypes.array.isRequired,
  isEventsLoading: PropTypes.bool.isRequired,
  isTemplatesLoading: PropTypes.bool.isRequired,
};

export default WebhookForm;
