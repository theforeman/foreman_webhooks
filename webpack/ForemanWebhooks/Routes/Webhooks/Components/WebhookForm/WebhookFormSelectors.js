import { STATUS } from 'foremanReact/constants';
import { deepPropsToCamelCase } from 'foremanReact/common/helpers';
import {
  selectAPIStatus,
  selectAPIResponse,
} from 'foremanReact/redux/API/APISelectors';

import {
  WEBHOOK_TEMPLATES_API_REQUEST_KEY,
  WEBHOOK_EVENTS_API_REQUEST_KEY,
} from '../../constants';

// Webhook templates selectors
const selectWebhookFormTemplatesResponse = state =>
  deepPropsToCamelCase(
    selectAPIResponse(state, WEBHOOK_TEMPLATES_API_REQUEST_KEY)
  );

const selectWebhookFormTemplatesStatus = state =>
  selectAPIStatus(state, WEBHOOK_TEMPLATES_API_REQUEST_KEY);

export const selectTemplatesHasError = state =>
  selectWebhookFormTemplatesStatus(state) === STATUS.ERROR;

export const selectTemplatesIsLoading = state => {
  const status = selectWebhookFormTemplatesStatus(state);
  return !status || status === STATUS.PENDING;
};

export const selectWebhookTemplates = state => {
  if (selectTemplatesHasError(state) || selectTemplatesIsLoading(state))
    return [];

  return selectWebhookFormTemplatesResponse(state).results;
};

// Webhook events selectors
const selectWebhookFormEventsResponse = state =>
  deepPropsToCamelCase(
    selectAPIResponse(state, WEBHOOK_EVENTS_API_REQUEST_KEY)
  );

const selectWebhookFormEventsStatus = state =>
  selectAPIStatus(state, WEBHOOK_EVENTS_API_REQUEST_KEY);

export const selectEventsHasError = state =>
  selectWebhookFormEventsStatus(state) === STATUS.ERROR;

export const selectEventsIsLoading = state => {
  const status = selectWebhookFormEventsStatus(state);
  return !status || status === STATUS.PENDING;
};

export const selectWebhookEvents = state => {
  if (selectEventsHasError(state) || selectEventsIsLoading(state)) return [];

  return selectWebhookFormEventsResponse(state);
};
