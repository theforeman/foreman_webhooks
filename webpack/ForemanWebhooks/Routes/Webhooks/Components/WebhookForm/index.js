import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { startCase, camelCase } from 'lodash';

import { foremanUrl } from 'foremanReact/common/helpers';
import { get } from 'foremanReact/redux/API';

import WebhookForm from './WebhookForm';

import {
  selectWebhookTemplates,
  selectWebhookEvents,
  selectTemplatesIsLoading,
  selectEventsIsLoading,
} from './WebhookFormSelectors';

import {
  WEBHOOK_TEMPLATES_API_REQUEST_KEY,
  WEBHOOK_TEMPLATES_API_PATH,
  WEBHOOK_EVENTS_API_REQUEST_KEY,
  WEBHOOKS_API_PLAIN_PATH,
} from '../../constants';

const params = { page: 1, search: 'snippet = false', per_page: 'all' };

const ConnectedWebhookForm = ({ onCancel, handleSubmit, initialValues }) => {
  const dispatch = useDispatch();

  const templates = useSelector(selectWebhookTemplates);
  const availableEvents = useSelector(selectWebhookEvents).map(e => ({
    value: `${e}.event.foreman`,
    label: startCase(camelCase(e)),
  }));
  const isTemplatesLoading = useSelector(selectTemplatesIsLoading);
  const isEventsLoading = useSelector(selectEventsIsLoading);

  useEffect(() => {
    dispatch(
      get({
        key: WEBHOOK_TEMPLATES_API_REQUEST_KEY,
        url: foremanUrl(WEBHOOK_TEMPLATES_API_PATH),
        params,
      })
    );
    dispatch(
      get({
        key: WEBHOOK_EVENTS_API_REQUEST_KEY,
        url: foremanUrl(`${WEBHOOKS_API_PLAIN_PATH}/events`),
      })
    );
  }, [dispatch]);

  return (
    <WebhookForm
      templates={templates}
      availableEvents={availableEvents}
      onCancel={onCancel}
      handleSubmit={handleSubmit}
      initialValues={initialValues}
      isTemplatesLoading={isTemplatesLoading}
      isEventsLoading={isEventsLoading}
    />
  );
};

ConnectedWebhookForm.propTypes = {
  onCancel: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.object.isRequired,
};

export default ConnectedWebhookForm;
