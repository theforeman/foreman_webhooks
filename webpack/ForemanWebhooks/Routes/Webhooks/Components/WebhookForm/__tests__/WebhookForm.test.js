import { testComponentSnapshotsWithFixtures } from '@theforeman/test';

import WebhookForm from '../WebhookForm';

const props = {
  handleSubmit: jest.fn(),
  onCancel: jest.fn(),
  onSuccess: jest.fn(),
  templates: [
    { id: 204, name: 'default template' },
    { id: 205, name: 'default template clone' },
  ],
  availableEvents: ['host_created'],
  isEventsLoading: false,
  isTemplatesLoading: false,
};

const fixtures = {
  'should render for new page': {
    ...props,
    initialValues: {
      http_method: 'POST',
      enabled: true,
      verify_ssl: true,
      http_content_type: 'application/json',
      event: 'host_created.event.foreman',
    },
  },
  'should render for edit page': {
    ...props,
    initialValues: {
      id: 54,
      http_method: 'PUT',
      enabled: true,
      verify_ssl: true,
      http_content_type: 'application/json',
      event: 'host_created.event.foreman',
      name: 'first webhook',
      target_url: 'https://foreman.example.com',
      user: undefined,
      password: undefined,
      webhook_template_id: 204,
      ssl_ca_certs: undefined,
    },
  },
};

describe('WebhookForm', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(WebhookForm, fixtures));
});
