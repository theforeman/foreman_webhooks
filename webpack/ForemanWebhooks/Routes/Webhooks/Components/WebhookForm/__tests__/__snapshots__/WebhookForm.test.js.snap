// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebhookForm rendering should render for edit page 1`] = `
<ForemanForm
  enableReinitialize={true}
  initialValues={
    Object {
      "enabled": true,
      "event": "host_created.event.foreman",
      "http_content_type": "application/json",
      "http_method": "PUT",
      "id": 54,
      "name": "first webhook",
      "password": undefined,
      "ssl_ca_certs": undefined,
      "target_url": "https://foreman.example.com",
      "user": undefined,
      "verify_ssl": true,
      "webhook_template_id": 204,
    }
  }
  item="Webhook"
  onCancel={[MockFunction]}
  onSubmit={[MockFunction]}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "webhook_template_id",
        "event",
        "http_method",
        "target_url",
        "name",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "fields": Object {
        "event": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "http_method": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "name": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "target_url": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "webhook_template_id": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
      "type": "object",
    }
  }
>
  <WebhookFormTabs
    activeTab={0}
    availableEvents={
      Array [
        "host_created",
      ]
    }
    disabled={false}
    formProps={Object {}}
    handleTabClick={[Function]}
    httpMethods={
      Array [
        Object {
          "label": "POST",
          "value": "POST",
        },
        Object {
          "label": "GET",
          "value": "GET",
        },
        Object {
          "label": "PUT",
          "value": "PUT",
        },
        Object {
          "label": "DELETE",
          "value": "DELETE",
        },
        Object {
          "label": "PATCH",
          "value": "PATCH",
        },
      ]
    }
    isEventsLoading={false}
    isTemplatesLoading={false}
    webhookTemplates={
      Array [
        Object {
          "label": "default template",
          "value": 204,
        },
        Object {
          "label": "default template clone",
          "value": 205,
        },
      ]
    }
  />
</ForemanForm>
`;

exports[`WebhookForm rendering should render for new page 1`] = `
<ForemanForm
  enableReinitialize={true}
  initialValues={
    Object {
      "enabled": true,
      "event": "host_created.event.foreman",
      "http_content_type": "application/json",
      "http_method": "POST",
      "verify_ssl": true,
    }
  }
  item="Webhook"
  onCancel={[MockFunction]}
  onSubmit={[MockFunction]}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "webhook_template_id",
        "event",
        "http_method",
        "target_url",
        "name",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "fields": Object {
        "event": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "http_method": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "name": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "target_url": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "webhook_template_id": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
      "type": "object",
    }
  }
>
  <WebhookFormTabs
    activeTab={0}
    availableEvents={
      Array [
        "host_created",
      ]
    }
    disabled={false}
    formProps={Object {}}
    handleTabClick={[Function]}
    httpMethods={
      Array [
        Object {
          "label": "POST",
          "value": "POST",
        },
        Object {
          "label": "GET",
          "value": "GET",
        },
        Object {
          "label": "PUT",
          "value": "PUT",
        },
        Object {
          "label": "DELETE",
          "value": "DELETE",
        },
        Object {
          "label": "PATCH",
          "value": "PATCH",
        },
      ]
    }
    isEventsLoading={false}
    isTemplatesLoading={false}
    webhookTemplates={
      Array [
        Object {
          "label": "default template",
          "value": 204,
        },
        Object {
          "label": "default template clone",
          "value": 205,
        },
      ]
    }
  />
</ForemanForm>
`;
